aip/general/0129.md

```diff
@@ -0,0 +1,71 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-modified Values and Defaults
+
+### State-driven clients (aka Convergent clients)
+Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.
+
+### Permadiff
+Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.
+
+## Guidance
+
+### Server Side Defaults
+There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.
+
+#### Dynamic Defaults
+When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with a dynamic default ***must*** be expressed as two fields in the API:
+* a mutable field that can be optionally set by the user
+* an immutable `OUTPUT_ONLY` field that records the final value decided on by the API service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string provisioned_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Static Defaults
+Service ***may*** choose to designate a value to be an equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version as specified in [AIP-180](aip-180). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default specifies machine-readable documentation of the field's behavior, allowing clients to understand default behavior and the meaning of an empty field.
```

Comment:
could we have a decision tree (it doesn't have to be a diagram, written is fine) about when an API author should opt to use a static default vs a dynamic default?

Putting the guidance of when to use the value in the sub-section forces the reader to read the whole document before knowing which section is relevant.